eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala

```diff
@@ -295,29 +318,112 @@ class PendingChannelsRateLimiterSpec extends ScalaTestWithActorTestKit(ConfigFac
+    limiter ! PendingChannelsRateLimiter.AddOrRejectChannel(probe.ref, peer, channelId1)
+    router.expectMessageType[GetNode].replyTo ! UnknownNode(peer)
+    probe.expectMessage(PendingChannelsRateLimiter.AcceptOpenChannel)
+
+    // peer makes second pending channel open requests
+    limiter ! PendingChannelsRateLimiter.AddOrRejectChannel(probe.ref, peer, channelId2)
+    router.expectMessageType[GetNode].replyTo ! UnknownNode(peer)
+    probe.expectMessage(PendingChannelsRateLimiter.AcceptOpenChannel)
+
+    // private channels are now at the limit
+    limiter ! PendingChannelsRateLimiter.AddOrRejectChannel(probe.ref, randomKey().publicKey, randomBytes32())
+    router.expectMessageType[GetNode].replyTo ! UnknownNode(randomKey().publicKey)
+    probe.expectMessage(PendingChannelsRateLimiter.ChannelRateLimited)
+
+    // when the first pending channel request is confirmed, the first tracked private channel is removed
+    // AND the peer becomes public, but still has a tracked channel request as a private node
+    system.eventStream ! Publish(ChannelOpened(null, peer, channelId1))
+    eventually {
+      limiter ! PendingChannelsRateLimiter.CountOpenChannelRequests(requests.ref, publicPeers = false)
+      requests.expectMessage(1)
+    }
+
+    // the third pending channel request from peer is accepted as a public peer
+    limiter ! PendingChannelsRateLimiter.AddOrRejectChannel(probe.ref, peer, channelId3)
+    router.expectMessageType[GetNode].replyTo ! PublicNode(announcement(peer), 1, 1 sat)
+    probe.expectMessage(PendingChannelsRateLimiter.AcceptOpenChannel)
+
+    // the private pending channel request from peer receives a new channel id
+    system.eventStream ! Publish(ChannelIdAssigned(null, peer, channelId1, randomBytes32()))
+
+    // the private channel request from peer is aborted and removed from the tracker
+    system.eventStream ! Publish(ChannelAborted(null, peer, channelId2))
+
+    // the first two private pending channel open requests are now removed from the tracker
+    eventually {
+      limiter ! PendingChannelsRateLimiter.CountOpenChannelRequests(requests.ref, publicPeers = false)
+      requests.expectMessage(0)
+    }
+
+    // the third pending channel request from peer is still tracked as a public peer
+    eventually {
+      limiter ! PendingChannelsRateLimiter.CountOpenChannelRequests(requests.ref, publicPeers = true)
+      requests.expectMessage(1)
+    }
+  }
+
+  test("receive events during the restore") { f =>
+    import f._
+
+    val events = Seq[ChannelEvent](
+      ChannelIdAssigned(null, peersAtLimit.head, channelIdAtLimit1, newChannelId1),
+      ChannelAborted(null, peersBelowLimit.last, channelIdBelowLimit2),
+      ChannelOpened(null, privatePeers.head, channelIdPrivate1),
+      ChannelOpened(null, peersAtLimit.head, newChannelId1))
+    val limiter = testKit.spawn(PendingChannelsRateLimiter(nodeParams, router.ref, allChannels))
+
+    // trigger the limiter to start the restore
```

Comment:
What does that mean? The limiter is already in the `restoring` state, isn't it? You're rather testing that while we're restoring, we reject channel requests?

Reply:
Fixed in [61e3c31](https://github.com/ACINQ/eclair/pull/2687/commits/61e3c3170cc9c62623fc52583c1ea6cd62c23d6f). I need a way to make sure the `limiter` has been created an subscribed to events before I publish the test events. What I had before was a dirty hack, this fix makes it cleaner/clearer.